fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/CustomDamageHandler.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1;
+
+import java.util.function.Consumer;
+
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ItemStack;
+
+/**
+ * Allows an item to run custom logic when {@link ItemStack#damage(int, LivingEntity, Consumer)} is called.
+ * This is useful for items that, for example, may drain durability from some other source before damaging
+ * the stack itself.
+ *
+ * <p>Custom damage handlers can be set with {@link FabricItemSettings#customDamage}.
+ */
+public interface CustomDamageHandler {
```

Comment:
I don't think it's possible, but just in case can this be turned into a functional interface so that the actual implementation of making a new CustomDamageHandler is a lot neater? Ideally I'd love to just be able to pass a lambda into the item settings method, but that might not be doable.

Reply:
This is effectively a functional interface but without the marker.

Reply:
Not really required, but for these kinds of interfaces it is advised to throw the `@FunctionalInterface` in

Reply:
Yeah but it uses a generic type parameter meaning it can't be expressed as a lambda. I tested an example of it in an IDE and it threw an error because the interface method was generic. Apparently this is just a limitation of java right now. The only way to allow it to be expressed as a lambda would be to remove the generic type parameter which sounds like it might end up being the plan.

Reply:
But yeah we should also add `@FunctionalInterface` to it